zerver/management/commands/runtusd.py

```diff
@@ -0,0 +1,57 @@
+import os
+from typing import Any
+from urllib.parse import parse_qs, urlencode, urlsplit, urlunsplit
+
+from django.conf import settings
+from django.core.management.base import BaseCommand, CommandParser
+
+from scripts.lib.zulip_tools import run
+
+
+def set_query_parameter(url: str, param_name: str, param_value: str) -> str:
+    scheme, netloc, path, query_string, fragment = urlsplit(url)
+    query_params = parse_qs(query_string)
+    query_params[param_name] = [param_value]
+    new_query_string = urlencode(query_params, doseq=True)
+
+    return urlunsplit((scheme, netloc, path, new_query_string, fragment))
+
+
+class Command(BaseCommand):
+    help = """Starts the TusD Server"""
+
+    def add_arguments(self, parser: CommandParser) -> None:
+        parser.add_argument("port", help="[Port to bind HTTP server to]", type=int)
+        parser.add_argument(
+            "hooks_http", help="[An HTTP endpoint to which hook events will be sent to]"
+        )
+
+    def handle(self, *args: Any, **options: Any) -> None:
+        port = options["port"]
+        hooks_http = options["hooks_http"]
+        hooks_http = set_query_parameter(hooks_http, "secret", settings.SHARED_SECRET)
```

Comment:
This isn't great, for two reasons -- first, POST requests which have a body as well as query parameters are mildly unusual.  Luckily, our `REQ` accessor handles it, by letting POST values override any query parameters.

But more importantly, query parameters are logged in request logs and the like, and are accessible to any intermediate proxies between the client and server.  This is _mostly_ OK in this context, because this request only over is over http to localhost, and so should only ever be able to end up in localhost's logs, which if you have access to you probably have access to the secrets on the server ... but it's still not good practice.

Does TUS provide any other ways to sign its hook requests?

Reply:
Yes, I tried to send secret in POST body, but they don't provide any way to do so. I will open a issue regarding this in tusd.

Reply:
That canonical way to do this is via HMAC, which means the secret doesn't actually go into the request at all.  See [Stripe](https://stripe.com/docs/webhooks/signatures) and [GitHub](https://docs.github.com/en/webhooks-and-events/webhooks/securing-your-webhooks) for prior art, though those are both looking at it from the point of view of the _consumer_ of the webhook (i.e. Zulip in this case) not the _caller_ (i.e. `tusd`)

Reply:
Thanks for suggestion. This is the best we can do for now, I will implement a new decorator or update existing one.

Reply:
There is no message to encrypt so I encrypted the SECRET using SECRET as key. This shouldn't be a problem, what are your thoughts ?

Reply:
I think it might make sense to try extracting an `hmac` based auth mechanism for inter-process communication as a new prep PR -- it's the kind of thing that's security-sensitive and would be nice to have in its own relatively simple PR. (The auth mechanism can be only called by unit tests in that PR).

Reply:
Okay, Makes sense. @timabbott do you agree with the approach, using SHARED_SECRET as message and key both?